src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -249,11 +249,14 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
 
         private static MethodInfo FirstImplementsIEquatableOfSecond(Type first, Type second)
         {
+            var pair = new KeyValuePair<Type, MethodInfo>();
+
             foreach (var xEquatableArgument in GetEquatableGenericArguments(first))
                 if (xEquatableArgument.Key.IsAssignableFrom(second))
-                    return xEquatableArgument.Value;
+                    if (pair.Key == null || pair.Key.IsAssignableFrom(xEquatableArgument.Key))
```

Comment:
The order of the return values is the order in which they are found by `Type.GetInterfaces`.

The Microsoft docs say

> The GetInterfaces method does not return interfaces in a particular order, such as alphabetical or declaration order. Your code must not depend on the order in which interfaces are returned, because that order varies.

That's pretty much par for the course. NUnit can never depend on the order of finding things by reflection. Even if it's constant for a give runtime, it will vary when we use others. So we would have to descend the inheritance hierarchy in order to provide a guaranteed order. We do this in other places, for example, when we are finding SetUp methods. That said, this is an improvement based on the tests that are passing.

@lundmikkel Do you want to take this further before we merge or should we go ahead with what you have done and do more work on it at another time? I'm OK either way.